api/src/test/java/org/pmiops/workbench/api/WorkspacesControllerTest.java

```diff
@@ -2318,4 +2324,179 @@ public void testGetWorkspacesDoesNotGetsPublishedIfReader() {
+  }
+
+  @Test
+  public void testNotebookLockingMetadata() {
+    final String lastLockedUser = LOGGED_IN_USER_EMAIL;
+    final Long lockExpirationTime = Instant.now().plus(Duration.ofMinutes(1)).toEpochMilli();
+
+    final Map<String, String> gcsMetadata =
+        new HashMap<String, String>() {
+          {
+            put("lockExpirationTime", lockExpirationTime.toString());
+            put(
+                "lastLockedBy",
+                WorkspacesController.notebookLockingEmailHash(BUCKET_NAME, lastLockedUser));
+            put("extraMetadata", "is not a problem");
+          }
+        };
+
+    // I can see that I have locked it myself, and when
+
+    final NotebookLockingMetadataResponse expectedResponse =
+        new NotebookLockingMetadataResponse()
+            .lockExpirationTime(lockExpirationTime)
+            .lastLockedBy(lastLockedUser);
+
+    assertNotebookLockingMetadata(gcsMetadata, expectedResponse, fcWorkspaceAcl);
+  }
+
+  @Test
+  public void testNotebookLockingMetadataKnownUser() {
+    final String readerOnMyWorkspace = "some-reader@fake-research-aou.org";
+
+    WorkspaceACL workspaceACL =
+        createWorkspaceACL(
+            new JSONObject()
+                .put(
+                    currentUser.getEmail(),
+                    new JSONObject()
+                        .put("accessLevel", "OWNER")
+                        .put("canCompute", true)
+                        .put("canShare", true))
+                .put(
+                    readerOnMyWorkspace,
+                    new JSONObject()
+                        .put("accessLevel", "READER")
+                        .put("canCompute", true)
+                        .put("canShare", true)));
+
+    final String lastLockedUser = readerOnMyWorkspace;
+    final Long lockExpirationTime = Instant.now().plus(Duration.ofMinutes(1)).toEpochMilli();
+
+    final Map<String, String> gcsMetadata =
+        new HashMap<String, String>() {
+          {
+            put("lockExpirationTime", lockExpirationTime.toString());
+            put(
+                "lastLockedBy",
+                WorkspacesController.notebookLockingEmailHash(BUCKET_NAME, lastLockedUser));
+            put("extraMetadata", "is not a problem");
+          }
+        };
+
+    // I'm the owner so I can see readers on my workspace
+
+    final NotebookLockingMetadataResponse expectedResponse =
+        new NotebookLockingMetadataResponse()
+            .lockExpirationTime(lockExpirationTime)
+            .lastLockedBy(readerOnMyWorkspace);
+
+    assertNotebookLockingMetadata(gcsMetadata, expectedResponse, workspaceACL);
+  }
+
+  @Test
+  public void testNotebookLockingMetadataUnknownUser() {
+    final String lastLockedUser = "a-stranger@fake-research-aou.org";
+    final Long lockExpirationTime = Instant.now().plus(Duration.ofMinutes(1)).toEpochMilli();
+
+    final Map<String, String> gcsMetadata =
+        new HashMap<String, String>() {
+          {
```

Comment:
This is really interesting - I've never seen this pattern used. I generally use an immutable map builder where I can... Does this anonymous class have an unfavorable side-effects? e.g. performance overhead or debug implications?

Reply:
ImmutableMap.Builder looks better so I updated it to use that.

I should really have a look around Guava to see what else I'm missing.